python/nav/web/auth.py

```diff
@@ -26,56 +26,111 @@
 
 
 def authenticate(username, password):
-    '''Authenticate username and password against database.
+    """
+    Authenticate username and password
+
+    First try LDAP, if available. Then fall back to Account.
+
     Returns account object if user was authenticated, else None.
-    '''
-    # FIXME Log stuff?
-    auth = False
-    account = None
+    """
+    account = authenticate_ldap(username, password)
+    if account:
+        return account
+    if account is False:  # An LDAP user is preferred over a db user
+        return None
+
+    account = authenticate_account(username, password)
+    if account:
+        return account
+
+    # Not authenticated
+    return None
+
+
+def authenticate_ldap(username=None, password=None):
+    """
+    Authenticate username and password against LDAP, if available
+
+    Returns account object if user was authenticated, False if the user is
+    invlid or otherwise is not allowed to log in, else None.
+    """
+    if not username and not password:
+        return None
+
+    if not ldapauth.available:
+        return None
+
+    try:
+        ldapuser = ldapauth.authenticate(username, password)
+    except ldapauth.NoAnswerError:
+        # LDAP unreachable, fallback
+        return None
+
+    try:
+        account = Account.objects.get(login__iexact=username)
+    except Account.DoesNotExist:
+        # Store the ldapuser in the database and return the new account
+        if ldapuser:
+            account = Account(
+                login=ldapuser.username,
+                name=ldapuser.get_real_name(),
+                ext_sync='ldap'
+            )
+            account.set_password(password)
+            account.save()
+            _logger.info("Created user %s from LDAP", account.login)
+            return account
+
+    # From this point on we have an existing Account
+
+    if ldapuser is False and account.ext_sync == 'ldap':
+        # This account is controlled by LDAP, and LDAP cannot
+        # authenticate the user
+        _logger.warn('Local database is out of sync with LDAP for user {}: '
+                    'either the password is wrong or the user is no '
+                    'longer registered in LDAP'.format(username))
+        return False
+
+    # From this point on we have an authenticated LDAPUser
+
+    # Bail out! Potentially evil user
+    if account.locked:
+        _logger.info("Locked ldap user %s tried to log in", account.login)
+        # Needs auditlog
+        return False
+
+    # Sync password from ldap to local db
+    if not account.check_password(password):
+        account.set_password(password)
+        _logger.info("Synced user %s's password from LDAP", account.login)
```

Comment:
This logic has at least one **glaring security problem**:

When LDAP is enabled, any non-LDAP-synced user (such as **the admin account**) will be able to log in using **any password they want**, as this will reset their password in the database to the password entered at login time.

Reply:
I thought that bit of code was dead. I wonder of the squash/rebase brought it back.

Having worked on the REMOTE_USER feature, I don't think this refactor is needed at all. The PR could be closed. But: didn't you wind up making a diagram of how the auth flow goes? That'd be the most valuable part of all of this I think, and it should be appended to the developers' documentation.

Reply:
Yes, the diagram was posted back in the #1888  that preceded this. I'll have a closer look at the REMOTE_USER PR ASAP and we'll se how it goes. These are sensitive codepaths to be meddling with :)